x-pack/platform/plugins/shared/streams/server/lib/telemetry/stats/streams_usage_collector.ts

```diff
@@ -0,0 +1,300 @@
+) {
+  return async function fetchStreamsUsageStats({
+    esClient,
+  }: {
+    esClient: any;
+  }): Promise<StreamsStatsTelemetry> {
+    try {
+      const streamsClient = await getStreamsClient();
+
+      const [allStreamsData, significantEventsMetrics, ruleExecutionMetrics] = await Promise.all([
+        streamsClient.listStreams(),
+        fetchSignificantEventsMetrics(esClient),
+        fetchRuleExecutionMetrics(esClient),
+      ]);
+
+      const { classicStreamsMetrics, wiredStreamsCount, streamsCache } =
+        processStreamsData(allStreamsData);
+
+      // Pass the cache to avoid additional ES calls
+      const categorizedSignificantEventsMetrics = await categorizeSignificantEvents(
+        significantEventsMetrics,
+        streamsCache
+      );
+
+      return {
+        classic_streams: classicStreamsMetrics,
+        wired_streams: { count: wiredStreamsCount },
+        significant_events: {
+          ...categorizedSignificantEventsMetrics,
+          ...ruleExecutionMetrics,
+        },
+      };
+    } catch (error) {
+      logger.error('Failed to collect Streams telemetry data', error);
+      throw error;
+    }
+  };
+
+  function processStreamsData(streamDefinitions: Streams.all.Definition[]) {
+    let changedCount = 0;
+    let withProcessingCount = 0;
+    let withFieldsCount = 0;
+    let withChangedRetentionCount = 0;
+    let wiredCount = 0;
+
+    const streamsCache = new Map();
+
+    for (const definition of streamDefinitions) {
+      if (Streams.WiredStream.Definition.is(definition)) {
+        wiredCount++;
+        streamsCache.set(definition.name, definition);
+      } else if (Streams.ClassicStream.Definition.is(definition)) {
+        // Presence of a classic stream in the streams client implies it has been stored/changed
+        changedCount++;
+
+        if (hasProcessingSteps(definition)) {
+          withProcessingCount++;
+        }
+
+        if (hasFieldOverrides(definition)) {
+          withFieldsCount++;
+        }
+
+        if (hasChangedRetention(definition.ingest.lifecycle)) {
+          withChangedRetentionCount++;
+        }
+
+        streamsCache.set(definition.name, definition);
+      }
+    }
+
+    return {
+      classicStreamsMetrics: {
+        changed_count: changedCount,
+        with_processing_count: withProcessingCount,
+        with_fields_count: withFieldsCount,
+        with_changed_retention_count: withChangedRetentionCount,
+      },
+      wiredStreamsCount: wiredCount,
+      streamsCache, // Return the cache
+    };
+  }
+
+  async function fetchSignificantEventsMetrics(esClient: any) {
+    const [rulesCount, eventsData] = await Promise.all([
+      fetchSignificantEventRulesCount(esClient),
+      fetchSignificantEventsData(esClient),
+    ]);
+
+    return {
+      rules_count: rulesCount,
+      ...eventsData,
+    };
+  }
+
+  async function fetchSignificantEventRulesCount(esClient: any) {
+    const rulesCountResponse = await esClient.search({
```

Comment:
@dgieselaar is this is a good way to get telemetry on the rules? I don't like how we are reaching into internal indices that are not maintained by this plugin and might change later on.

Reply:
I think the alerting plugin by itself collects rule-count-by-type telemetry